File: src/Infernu/Builtins/Date.hs
{-# LANGUAGE TupleSections #-}
module Infernu.Builtins.Date
       (dateRowType)
       where

import           Control.Monad         (foldM)

import           Infernu.Builtins.Util
import           Infernu.InferState    (Infer)
import           Infernu.Prelude
import           Infernu.Types

dateFunc :: Type -> Type
dateFunc = funcN [date]

getNumFromDate :: TScheme Type
getNumFromDate = ty $ dateFunc number

getNumFuncs :: [String]
getNumFuncs = [ "getDate"
              , "getDay"
              , "getFullYear"
              , "getHours"
              , "getMilliseconds"
              , "getMinutes"
              , "getMonth"
              , "getSeconds"
              , "getTime"
              , "getTimezoneOffset"
              , "getUTCDate"
              , "getUTCDay"
              , "getUTCFullYear"
              , "getUTCHours"
              , "getUTCMilliseconds"
              , "getUTCMinutes"
              , "getUTCMonth"
              , "getUTCSeconds"
              , "valueOf"
              ]

setNumOnDate :: TScheme Type
setNumOnDate = ty $ funcN [date, number] number

-- Note: many of these functions accept multiple optional parameters, but there are other ways to
-- achieve the same result. Leaving these optional parameters out.
setNumFuncs :: [String]
setNumFuncs =
    [ "setDate"
    , "setFullYear"
    , "setHours"
    , "setMilliseconds"
    , "setMinutes"
    , "setMonth"
    , "setSeconds"
    , "setTime"
    , "setUTCDate"
    , "setUTCFullYear"
    , "setUTCHours"
    , "setUTCMilliseconds"
    , "setUTCMinutes"
    , "setUTCMonth"
    , "setUTCSeconds"
      --               , "setYear"
    ]

getStringFuncType :: TScheme (Fix FType)
getStringFuncType = ty $ funcN [date] string

getStringFuncs :: [String]
getStringFuncs =
    [ "toDateString"
--    , "toGMTString"
    , "toISOString"
    , "toJSON"
--    , "toLocaleFormat"
--    , "toSource"
    , "toString"
    , "toTimeString"
    , "toUTCString"
    ]
    -- , "toLocaleDateString"
    -- , "toLocaleString"
    -- , "toLocaleTimeString"

dateProps :: [(String, TypeScheme)]
dateProps =
    map (,getNumFromDate) getNumFuncs
    ++ map (,setNumOnDate) setNumFuncs
    ++ map (,getStringFuncType) getStringFuncs

dateRowType :: Infer (TRowList Type)
dateRowType = namedProps
  where namedProps = foldM addProp (TRowEnd Nothing) dateProps

File: test/valid/builtins/Date.js
var x = new Date();
var y = Date.now();
var z = Date.parse('2015-01-01');
var u = Date.UTC(2015,3,1,11,10,9,932);


var n = 1;

n = x.getDate();
n = x.getDay();
n = x.getFullYear();
n = x.getHours();
n = x.getMilliseconds();
n = x.getMinutes();
n = x.getMonth();
n = x.getSeconds();
n = x.getTime();
n = x.getTimezoneOffset();
n = x.getUTCDate();
n = x.getUTCDay();
n = x.getUTCFullYear();
n = x.getUTCHours();
n = x.getUTCMilliseconds();
n = x.getUTCMinutes();
n = x.getUTCMonth();
n = x.getUTCSeconds();
n = x.valueOf();

x.setDate(n);
x.setFullYear(n);
x.setHours(n);
x.setMilliseconds(n);
x.setMinutes(n);
x.setMonth(n);
x.setSeconds(n);
x.setTime(n);
x.setUTCDate(n);
x.setUTCFullYear(n);
x.setUTCHours(n);
x.setUTCMilliseconds(n);
x.setUTCMinutes(n);
x.setUTCMonth(n);
x.setUTCSeconds(n);
//x.setYear(n);

var s = '';
s = x.toDateString();
s = x.toISOString();
s = x.toJSON();
s = x.toString();
s = x.toTimeString();
s = x.toUTCString();

